src/control/common/pci_utils.go

```diff
@@ -37,3 +48,302 @@ func ParsePCIAddress(addr string) (dom, bus, dev, fun uint64, err error) {
+func NewPCIAddress(addr string) (*PCIAddress, error) {
+	dom, bus, dev, fun, err := parsePCIAddress(addr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to parse %q", addr)
+	}
+
+	pa := &PCIAddress{
+		Bus:      fmt.Sprintf("%02x", bus),
+		Device:   fmt.Sprintf("%02x", dev),
+		Function: fmt.Sprintf("%01x", fun),
+	}
+
+	if dom == 0 {
+		pa.Domain = "0000"
+	} else {
+		pa.Domain = fmt.Sprintf("%x", dom)
+	}
+
+	return pa, nil
+}
+
+// PCIAddressSet represents a unique list of PCI addresses.
+type PCIAddressSet struct {
+	addrMap map[string]*PCIAddress
+}
+
+// Contains returns true if provided address is already in list.
+func (pas *PCIAddressSet) Contains(a *PCIAddress) bool {
+	if pas == nil {
+		return false
+	}
+
+	_, found := pas.addrMap[a.String()]
+	return found
+}
+
+func (pas *PCIAddressSet) add(a *PCIAddress) {
+	if pas.addrMap == nil {
+		pas.addrMap = make(map[string]*PCIAddress)
+	}
+
+	pas.addrMap[a.String()] = a
+}
+
+// Add adds PCI addresses to slice. Ignores duplicate addresses.
+func (pas *PCIAddressSet) Add(addrs ...*PCIAddress) error {
+	if pas == nil {
+		return errors.New("PCIAddressSet is nil")
+	}
+
+	for _, addr := range addrs {
+		pas.add(addr)
+	}
+
+	return nil
+}
+
+// AddStrings adds PCI addresses to slice from supplied strings. If any input string is not a valid PCI
+// address then return error and don't add any elements to slice. Ignores duplicateaddresses.
+func (pas *PCIAddressSet) AddStrings(addrs ...string) error {
+	if pas == nil {
+		return errors.New("PCIAddressSet is nil")
+	}
+
+	for _, addr := range addrs {
+		if addr == "" {
+			continue
+		}
+
+		a, err := NewPCIAddress(addr)
+		if err != nil {
+			return err
+		}
+
+		pas.add(a)
+	}
+
+	return nil
+}
+
+// Addresses returns sorted slice of PCI address type object references.
+func (pas *PCIAddressSet) Addresses() []*PCIAddress {
+	if pas == nil {
+		return nil
+	}
+
+	addrs := make([]*PCIAddress, 0, len(pas.addrMap))
+	for _, addr := range pas.addrMap {
+		addrs = append(addrs, addr)
+	}
+	sort.Slice(addrs, func(i, j int) bool { return addrs[i].LessThan(addrs[j]) })
+
+	return addrs
+}
+
+// Strings returns PCI addresses as slice of strings.
+func (pas *PCIAddressSet) Strings() []string {
+	if pas == nil {
+		return nil
+	}
+
+	addrs := make([]string, len(pas.addrMap))
+	for i, addr := range pas.Addresses() {
+		addrs[i] = addr.String()
+	}
+
+	return addrs
+}
+
+// Strings returns PCI addresses as string of joined space separated strings.
+func (pas *PCIAddressSet) String() string {
+	return strings.Join(pas.Strings(), bdevPciAddrSep)
+}
+
+// Len returns length of slice. Required by sort.Interface.
+func (pas *PCIAddressSet) Len() int {
+	if pas == nil {
+		return 0
+	}
+
+	return len(pas.addrMap)
+}
+
+// IsEmpty returns true if address set is empty.
+func (pas *PCIAddressSet) IsEmpty() bool {
+	return pas.Len() == 0
+}
+
+func hexStr2Int(s string) int64 {
+	i, err := strconv.ParseInt(s, 16, 64)
+	if err != nil {
+		panic(err)
```

Comment:
Is panic a bit harsh for here? I know if you have a bad token in the address you can't really recover from it but doesn't panic crash the system? Wouldn't it be better to try to provide a better error message about the specific token further up the stack?

Reply:
I think it's okay here, the current usage will only ever process addresses that have been previously validated. reworking gets really nasty/messy.